File: Data Structures & Algorithms/jump-game/submission-0.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int currTarget = n - 1;
        int prevTarget = currTarget;

        while (currTarget != 0) {
            for (int i = currTarget - 1; i >= 0; --i) {
                if (nums[i] + i >= currTarget) {
                    currTarget = i;
                    break;
                }
            }
            if (prevTarget == currTarget) return false;
            prevTarget = currTarget;
        }
        return true;
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp1 = [0] * len(nums)
        dp2 = [0] * len(nums)

        for i, val in enumerate(nums):
            if i == 0:
                dp1[0] = 0
            elif i == 1:
                dp1[i] = val
            else:
                dp1[i] = max(val + dp1[i - 2], dp1[i - 1])

        for i, val in enumerate(nums):
            if i == 0:
                dp2[0] = val
            elif i == 1 or i == len(nums) - 1:
                dp2[i] = dp2[i - 1]
            else:
                dp2[i] = max(val + dp2[i - 2], dp2[i - 1])

        return max(dp1[-1], dp2[-1])
        
        